File: src/Day18.hs

module Day18 (main) where

import Control.Applicative (many)
import Misc (check)
import ParE (Par,parse,separated,nl,alts,int,sp,lit)

main :: IO ()
main = do
  sam <- readFile "input/day18.input.sam"
  inp <- readFile "input/day18.input"
  let !_ = check 13632 $ part1 sam
  print ("day18, part1", check 75592527415659 $ part1 inp)
  let !_ = check 23340 $ part2 sam
  print ("day18, part2", check 360029542265462 $ part2 inp)

part1 :: String -> Integer
part1 s = sum (map eval (parse gram1 s))

part2 :: String -> Integer
part2 s = sum (map eval (parse gram2 s))

eval :: Exp -> Integer
eval = \case
  Num n -> n
  Add e1 e2 -> eval e1 + eval e2
  Mul e1 e2 -> eval e1 * eval e2

data Exp = Num Integer | Add Exp Exp | Mul Exp Exp
  deriving Show

gram1 :: Par [Exp]
gram1 = separated nl exp
  where
    atom :: Par Exp
    atom = alts [ (Num . fromIntegral) <$> int , bracketed exp ]

    exp :: Par Exp
    exp = do
      x <- atom
      xs <- many $ do
        sp
        f <- alts [ do lit '+'; pure Add
                  , do lit '*'; pure Mul ]
        sp
        x <- atom
        pure (f,x)

      pure $ foldl node x xs
        where node e1 (f,e2) = f e1 e2

gram2 :: Par [Exp]
gram2 = separated nl exp
  where
    atom :: Par Exp
    atom = alts [ (Num . fromIntegral) <$> int , bracketed exp ]

    thing :: Par Exp
    thing = do
      x <- atom
      xs <- many $ do
        sp
        f <- do lit '+'; pure Add
        sp
        x <- atom
        pure (f,x)

      pure $ foldl node x xs
        where node e1 (f,e2) = f e1 e2

    exp :: Par Exp
    exp = do
      x <- thing
      xs <- many $ do
        sp
        f <- do lit '*'; pure Mul
        sp
        x <- thing
        pure (f,x)

      pure $ foldl node x xs
        where node e1 (f,e2) = f e1 e2

bracketed :: Par a -> Par a
bracketed p = do
  lit '('
  x <- p
  lit ')'
  pure x

File: python/day18.py
#!/usr/bin/env python

class Num:
    def __init__(self, n):
        self.n = n
    def pretty(self):
        return str(self.n)
    def eval(self):
        return self.n

class Add:
    def __init__(self, a, b):
        self.a = a
        self.b = b
    def pretty(self):
        return "(" + self.a.pretty() + " + " + self.b.pretty() + ")"
    def eval(self):
        return self.a.eval() + self.b.eval()

class Mul:
    def __init__(self, a, b):
        self.a = a
        self.b = b
    def pretty(self):
        return "(" + self.a.pretty() + " * " + self.b.pretty() + ")"
    def eval(self):
        return self.a.eval() * self.b.eval()

#sam = Add(Num(100),Add(Num(20),Num(3)))
#print(sam.pretty())
#print(sam.eval())

def parse(s):
    max = len(s)
    pos = 0
    def get(who):
        nonlocal pos
        if pos < max:
            c = s[pos]
            pos = pos + 1
            #print("get("+who+"): s["+str(pos)+"] -> '"+c+"'")
            return c
        else:
            return None

    def put_back():
        nonlocal pos
        pos = pos - 1

    def digit():
        c = get("digit")
        if (c is None):
            return None
        if (c < '0' or c >'9'):
            put_back()
            return None
        return c;

    def numberAcc(acc):
        c = digit()
        if (c is None):
            return acc;
        else:
            return numberAcc(10*acc + int(c))

    def number():
        c = digit()
        if (c is None):
            return None
        return numberAcc(int(c))

    def err(msg):
        print(msg + " at " + str(pos) + ", got: '" + s[pos] + "'")

    def lit(x,who):
        c = get(who)
        if (c == x):
            return True
        else:
            put_back()
            return None

    def atom():
        if (lit('(',"open")):
            a = term()
            nibble()
            if (lit(')',"close")):
                return a
            else:
                err("atom() expected a matching close paren")
        else:
            n = number();
            if (n is None):
                err("atom() expected a number or bracketed expression")
            return Num(n)

    def nibble():
        if (lit(' ',"space") is None):
            return
        else:
            nibble()

    def op():
        if (lit('+',"plus") is None):
            if (lit('*',"star") is None):
                return None
            else:
                return Mul
        else:
            return Add

    def termAcc(a1):
        nibble()
        o = op()
        if (o is None):
            return a1
        nibble()
        a2 = atom()
        if (a2 is None):
             err("termAcc() expected an atom after op")
        return termAcc(o(a1,a2))

    def term():
        a1 = atom()
        return termAcc(a1)

    #print("s:"+s)
    #print("max="+str(max))
    a = term()
    #print("parse(): done, pos="+str(pos))
    return a


sam = open('input/day18.input.sam', 'r').read()
a = parse(sam)
#print(a.pretty())
print(a.eval())

inp = open('input/day18.input', 'r').read().splitlines()
res = sum([ parse(line).eval() for line in inp ])
assert (res == 75592527415659)
print ("day18,part1 : ", res)

